compiler/rustc_infer/src/infer/canonical/query_response.rs

```diff
@@ -248,20 +248,69 @@ impl<'tcx> InferCtxt<'tcx> {
     where
         R: Debug + TypeFoldable<TyCtxt<'tcx>>,
     {
-        let InferOk { value: result_args, mut obligations } = self
-            .query_response_instantiation_guess(
-                cause,
-                param_env,
-                original_values,
-                query_response,
-            )?;
+        let (result_args, opaques) =
+            self.query_response_instantiation_guess(cause, original_values, query_response);
+
+        let mut obligations = vec![];
+
+        let constraint_category = cause.to_constraint_category();
+
+        // Carry all newly resolved opaque types to the caller's scope
+        for &(a, b) in &opaques {
+            let a = instantiate_value(self.tcx, &result_args, a);
+            let b = instantiate_value(self.tcx, &result_args, b);
+            debug!(?a, ?b, "constrain opaque type");
+            if let ty::Alias(ty::Opaque, alias_ty) = b.kind()
+                && alias_ty.def_id == a.def_id.into()
+            {
+                // When an inference variable is registered as the hidden type of an opaque type,
+                // it may end up resolved to the opaque type itself.
+                // In that case, do not bubble up the opaque type, but instead just equate its generic params.
+                assert_eq!(a.args.len(), alias_ty.args.len());
+                for (a, b) in a.args.iter().zip(alias_ty.args) {
+                    match (a.unpack(), b.unpack()) {
+                        (GenericArgKind::Lifetime(v_o), GenericArgKind::Lifetime(v_r)) => {
+                            // To make `v_o = v_r`, we emit `v_o: v_r` and `v_r: v_o`.
+                            if v_o != v_r {
+                                output_query_region_constraints.outlives.push((
+                                    ty::OutlivesPredicate(v_o.into(), v_r),
+                                    constraint_category,
+                                ));
+                                output_query_region_constraints.outlives.push((
+                                    ty::OutlivesPredicate(v_r.into(), v_o),
+                                    constraint_category,
+                                ));
+                            }
+                        }
+                        // FIXME: actually equate types and consts by just registering a AliasRelate obligations for the opaque type
+                        // just like the new solver does. Needs AliasRelate support in the old solver.
+                        _ => {
+                            if a != b {
+                                span_bug!(
+                                    cause.span(),
+                                    "self-referential opaque type can only differ in lifetime parameters, but got {a} != {b}"
+                                )
+                            }
+                        }
```

Comment:
This is super hacky, but at least it's a high signal ICE instead of some hard-to-debug issues. We also can't work on the FIXME before we get rid of `DefiningAnchor`, so I think we should just roll with this for now and address it later.